HLD: Handling table cell state when scrolling

msmithNI opened the pull request:
# Pull Request

## 🤨 Rationale

HLD for #997 

## 👩‍💻 Implementation

N/A

## 🧪 Testing

N/A

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

atmgrifter00 commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,98 @@
+If we have a focused cell:
+
+-   If the cell has an action menu (note: in this case the focused element will be a `nimble-menu-item`), find the associated `nimble-menu-button`. If open, set `open` to `false`.
+-   Get the `TableColumn` instance for the column that the cell is in, and call the new API `onBeforeFocusedCellRecycled(cell: TableCell)` on it (if defined).
```
I guess I don't have the clearest picture of why this abstract method would be warranted (and ultimately, this feels like a bit of an awkward concern to hoist upon the column API). If I understand the flow of things, by the time you would have called this abstract method we will have already discovered the element that is the `document.activeElement` (and whether it is in a `nimble-table-cell`). At this point, couldn't we just dumbly call `blur()` on that element on scroll? What would this do for menu components that are in an open state? I would think for non-input components, this would result in a no-op, or at least nothing of consequence (I could be wrong here).

msmithNI replied:
If we find the leafmost active element (instead of stopping at `nimble-table-cell`), we'll stop at the `input` or `textarea` inside the `nimble-text-field`/`nimble-number-field`/`nimble-text-area`. Calling `blur()` on those should work, and be sufficient.
Calling `blur()` on a `nimble-menu-item` is fine (it doesn't throw an error and it does lose focus), but I saw some cases where the menu / menu button remained open after doing that. But I suppose we could still have custom code to close menus if needed directly in the virtualizer without adding the `onBeforeFocusedCellRecycled` API.

msmithNI replied:
Current plan is to not have a default implementation of `onBeforeFocusedCellRecycled` (still have a note about it in Open Issues if there's any disagreement on that).

The proposal @rajsite outlined above about firing a custom event on the cell (cell's content container specifically, based on our chat) is captured in the HLD as "Option 1" in the section about blurring cells. The original `onBeforeFocusedCellRecycled` proposal is Option 2.  

Added an Open Issue to choose between those 2 proposals.

atmgrifter00 commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,98 @@
+
+[#997: Table virtualization: Clear/Commit (and potentially save/restore) row/cell state when scrolling](https://github.com/ni/nimble/issues/997)
+
+## Implementation / Design
```
Would it be helpful for us to point to prior art on how various tables handle these use cases?

jattasNI replied:
My initial position before reading the spec was "text selection should work as if the table isn't being virtuallized" but after reading your (excellent) in-depth research I am also coming around to the position that it may not be worth doing.

Another good next step would be to give some high-level estimates to Nate of the up-front work and the maintenance burden of this code. Then he can help guide us between the current recommendation in the spec vs more drastic measures like disabling selection on the table entirely.

msmithNI replied:
It's worth pointing out that the `nimble-table` text selection issues are present even without virtualization (as I said, I suspect they're more due to the table using shadow DOM).  
[Here is a Storybook link](https://60e89457a987cf003efc0a5b-pffjjtbdsl.chromatic.com/?path=/story/table--table) from before virtualization was added and it shows the same text selection behavior I described.

As Jesse alluded to, disabling text selection for table text cells (`user-select: none`) is another potential option. So the full set of options is probably:
- Do nothing (current behavior). User can select table cell text with difficulty, text remains selected after a scroll (and different text gets selected)
- Disable text selection for table text cells with `user-select: none` (therefore, there's no text selection that would be kept after a scroll)
- Detect and clear text selection on scroll as this HLD proposed.

jattasNI replied:
My preference order:
1. disable text selection. It sounds so inconsistent across browsers that I'm worried end users will get frustrated and clients will think it's buggy
2. do nothing.
3. detect and clear text selection on scroll. It seems expensive/fragile and the payoff isn't that high given the challenges selecting text in the first place

I think we should get opinions from the team but if there's much contention I'd say it's Nate's call to decide which option gives the best ROI

rajsite replied:
My vote is do nothing.

I think the basic text selection interaction stuff for single items or a single row is likely fixable (probably the fixes Molly mentioned). I'd be much more generally concerned if not and would want to understand what's going on and how that impacts our controls outside the table too.

I think the selected row scrolling isn't a big issue as discussed in stand-up (not a primary interaction, shouldn't harm functionality, easy to resolve for the user by clicking off to deselect text).

I think disabling all text selection is heavy handed, being able to just quickly copy a cell of data or select some text, right-click, and search google is handy.

I think the programmatic stuff for detecting selected text seems complex and likely fragile. It's known that shadow dom + selection apis have issues and likely to change over time, i.e. from a developer contributing a lot to web component apis in browser: https://twitter.com/rniwa_dev/status/1624193928659607552

msmithNI replied:
Molly just posted a PR with the fixes, [here's a Storybook link](https://60e89457a987cf003efc0a5b-miemcebjrx.chromatic.com/?path=/story/table--table&args=data:LargeDataSet).

It fixes the text selection issues in Chrome (doubleclicking any cell text selects that cell) and Safari (you can drag-select text in one cell / multiple cells / a whole row now, and doubleclicking text to select works).  
Firefox still selects a maximum of one cell of text, but it's easy to do so now.

jattasNI replied:
With Molly's fixes I'm fine with either "disable text selection" or "do nothing".

msmithNI replied:
Updated HLD to state that we won't detect/remove table text selection. Moved the potential implementation for that to Alternatives. Add "Disabling Text Selection" to the Alternatives section too.

rajsite commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,106 @@
+
+If we have a focused cell:
+
+-   If the cell has an action menu (note: in this case the focused element will be a `nimble-menu-item`), find the associated `nimble-menu-button`. If open, set `open` to `false`.
```
I think as written this is trying to make a generic implementation and that generic implementation will cover the action menu. I think we should just own the action menu within the table instead since the plugin api is not responsible for it.

msmithNI commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,106 @@
+
+If we have a focused cell:
+
+-   If the cell has an action menu (note: in this case the focused element will be a `nimble-menu-item`), find the associated `nimble-menu-button`. If open, set `open` to `false`.
```
Updated the HLD to have separate sections for "Blur Focused Controls in Cells" vs "Close Focused Action Menus".

In the action menu section, clarified that the table will handle that itself (in the Virtualizer class), it won't affect the TableColumn API.

msmithNI commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,106 @@
+
+**Planned API:**  
+New optional abstract method `onBeforeFocusedCellRecycled(cell: TableCell)` on the `TableColumn` abstract class.  
+New logic in Nimble Virtualizer class, `handleVirtualizerChange()` (called when the user scrolls), to find the active/focused cell.  
```
Correct. Updated the HLD.

The [prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-e2bfd4eda0a3c89f54b4e09624e6cfd5a68ea2f14c0e79f717eacce38ec5982bR120) shows the checks we'd need to decide if more work was warranted on a scroll:
- Focused control in a table cell: in that case `table.shadowRoot.activeElement` is a TableRow and we can continue to the cell from there, and trigger the appropriate method/logic on the cell that'll let the column implementation blur the focused control.
- A menu item from an action menu is focused: in this case `table.shadowRoot.activeElement` is null. However `document.activeElement` is a Nimble menu item. First we can check if `table.contains(menuItem)` before doing the rest of the logic (find the associated menu, find the associated cell, find the menu button, and close it).

msmithNI commented:
@rajsite updated the HLD with the proposal we talked about.
@jattasNI @mollykreis Milan floated another idea of raising an event on the cell/content container that can be handled from the column implementor's template, instead of the original method proposed (`onBeforeFocusedCellRecycled`).  
I reworked the HLD to cover both of those ideas, so let me know if you have a preference. I also moved all of the text selection info to "Alternate Implementations" since based on previous discussions, we're not planning on doing anything for text selection.

msmithNI commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,106 @@
+
+-   If the cell has an action menu (note: in this case the focused element will be a `nimble-menu-item`), find the associated `nimble-menu-button`. If open, set `open` to `false`.
+-   Get the `TableColumn` instance for the column that the cell is in, and call the new API `onBeforeFocusedCellRecycled(cell: TableCell)` on it (if defined).
+    -   General expected implementation, for a column type with editable controls in their `cellTemplate` would be to find the editable control with `querySelector()`, and close/commit it. i.e. `blur()` for `nimble-text-field`/`nimble-text-area`/`nimble-number-field` or other input controls, `open = false` for `nimble-menu-button`/`nimble-combobox`/etc.
```
Note: This is related to what's called "Option 1" in the HLD. I didn't investigate this idea more yet so keeping this comment thread open for now.

rajsite replied:
This came up in another PR: https://github.com/ni/nimble/pull/1092#discussion_r1124875059

I think what we are landing on is that we do need a better way to represent cell state for plugin authors. I think this HLD could cover the scope of having a way to track instance state correctly and how that instance state is addressed during scroll actions.

In that above linked issue I'm essentially advocating that the column plugin api change from having a cellTemplate and cellStyles that are stateless to the plugin author defining a component. What we are finding is that cellTemplate does have state from the view and is not stateless, i.e. if  something in it's template is focused, and we are also finding that column authors need a place to store state, i.e. the cell is hovered and I need to update my title.

In that scenario a plugin might look like:
```ts
// Column definition
import './my-column-view';
export class MyColumn extends TableColumn {
    public readonly viewTag = 'nimble-my-column-view';
}

// my-column-view
export class MyColumnView extends TableColumnView {
    public myMenu;

    onTableBlur () {
        this.myMenu.open = false;
    }
}

// template
const template = html`
        <nimble-menu ${ref('myMenu')}></nimble-menu>
`;

// TableColumnView
TableColumnView extends FoundationElement {
    @observable public celldata;
    @observable public columnConfig;
    abstract onTableBlur;
}
```

So now with the above we have a base class element the user is extending and we can require some api interface be implemented, ie onTableBlur, etc. It's part of the cell view api by being on the cell view element (where the weird part before was being a method on the column element which is completely separate from the view).

And we can make a clear indication that the user needs to implement that behavior to render in the table by making it abstract in the base class where with the event approach they need docs to know if that event changes / needs to be implemented.

rajsite replied:
Discussed with the team and @msmithNI is going to put together a prototype branch to sanity check. If looks good will merge into this HLD for how the table notifies the view about events to respond to (onBlur from this spec and others discussed were potentially an onRecycled hook).

msmithNI replied:
@rajsite @atmgrifter00 Incorporated this newer proposal into the HLD. Also took a first pass at updating the Table Column HLD with changes that are due to the custom-elements-in-cells proposal. It's possible other HLDs will need updates too (I didn't finish looking through all of them yet).

jattasNI commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,111 @@
+
+When the user scrolls the table, we will:
+
+-   'Blur' (Lose focus) from any active/focused control in a table cell. Generally this means that whatever text the user began to change in that control will be committed, not discarded.
```
In playing with your prototype, if I have a pending edit in one of the text fields and then start scrolling, the pending edit is discarded. That appears to contradict what's written here, but is that just a limitation of the prototype? I'm guessing we don't have any code that handles the blur and updates the data source?

msmithNI replied:
Correct, there's just no code in the prototype that updates the data source based on the edits.

rajsite replied:
Maybe we can phrase this one as the table will notify a column plugin that that they will lose focus. The recommended behavior for plugins is to commit changes when appropriate and blur.

msmithNI replied:
Updated wording in the HLD. The updated prototype example also no longer uses an editable text box, so the point no longer applies.

jattasNI commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,111 @@
+
+We will not:
+
+-   Detect and deselect selected table text on scroll. This means that if table text is selected and the user scrolls, text will still be selected (potentially different text). We don't consider that to be a primary table interaction, and it shouldn't harm functionality (it's easy to resolve for the user by clicking off to deselect text). It's also the same behavior as the current `sl-grid`/`smar ...
```
In the prototype, selected text within a text field is deselected. I'm guessing that's a side effect of calling blur, not something that will happen for non-input text? If so it might be good to clarify that in this description.

msmithNI replied:
Correct, that's a result of calling blur() so it'll only be for text in editable controls. I'll update the HLD.

msmithNI replied:
I revamped the prototype to use focusable-but-not-editable text. Hopefully that'll simplify things a bit, and remove the issues/questions related to editable controls and updating the table value that aren't directly related to this HLD.

jattasNI reviewed (approved):
I don't have an opinion on option 1 vs 2 but let me know if you need help making that decision.

atmgrifter00 commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,111 @@
+
+If we found a focused `TableCell`, we then want to `blur()` the control in it:  
+(**Open Question: Pick one of the following two approaches**)  
+**Option 1:**  
```
My vote is for this option. I far prefer to have an API semantically oriented around the notion of `blur`, even if what we have to document is that a column implementer must manually call `blur` for the proper behavior.

Avoiding semantics associated with cell recycling in our public API seems highly desirable.

rajsite replied:
+1 for blur semantics. +1 for template events for responding to table actions in column apis.

The table column api could be called onCellBlur instead of onBeforeFocusedCellRecycled in which case both the template event based approach and column callback approach have blur semantics.

But I prefer the template event based approach either way. It nicely encapsulates behavior in the template and keeps the column author api consistent as pretty dumb configuration objects.

msmithNI replied:
Current proposal is a new `onBeforeBlur()` method on the new `BaseCellElement` class (which is subclassed for each different column type). The state/behavior will all be in the new cell element, not in the column object.

rajsite commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,111 @@
+
+-   Decide between API Option 1 and Option 2 for blur-ing focused cell controls.
+-   Should we have a default, generic implementation of `onBeforeFocusedCellRecycled`? (It would look for known Nimble control types in the cell, and call the appropriate API on them, `blur()` or `open = false`)
+    -   Current we **do not** plan to have a default implementation. We think it's better for the column implementations to handle this, as they're the ones declaring the editable/blur-able controls in their templates.
```
+1

msmithNI replied:
Took out this open issue. Added a section in Alternate Implementations about the idea of having this default logic in the blur handler, but that we decided against it.

mollykreis commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,111 @@
+The table will handle this internally (in the `Virtualizer` class, via `handleVirtualizerChange()`), without affecting the TableColumn public API.
+
+We want to close the action menu via the associated `MenuButton`, which allows the rest of the table logic dealing with action menus to get called normally.
+In this case, `table.shadowRoot.activeElement` will be null (since the action menus are slotted in), but `document.activeElement` will be a Nimble `MenuItem`. (We can also doublecheck that `table.contains(document.activeElement)` before proceeding.) From the MenuItem, we can walk up the `parentElement` hierarchy to get to the Nimble `Menu`. Then we can walk up the `assignedSlot` hierarchy until w ...
```
Rather than having to walk the DOM tree & make assumptions about the focused element, an alternative is to use the information we already have in the table & table-row about which menu is open.

Specifically:
- the table has an observable property named `openActionMenuRecordId`
- the table row has `menuOpen` and `currentActionMenuColumn`

Without thinking through the exact code, my initial preference is to use the information we have already (we might find we want to make some slight changes to reset some values back to `undefined` when menus are closed) to programmatically close the open menu without walking the DOM.

msmithNI replied:
I took a stab at this in the [updated prototype](https://github.com/ni/nimble/compare/@ni/nimble-components_v18.6.1...table-cell-state-custom-elements-1#diff-e2bfd4eda0a3c89f54b4e09624e6cfd5a68ea2f14c0e79f717eacce38ec5982bR132), but the code still doesn't seem optimal.
- `openActionMenuRecordId` to `TableRow` - I ended up doing ``viewport.querySelector(`nimble-table-row[record-id="${this.table.openActionMenuRecordId}"]`)``, not sure if there's a more straightforward way
- From `currentActionMenuColumn` I don't think we can easily get to a `TableCell` - given a row, we can get all the cells (via DOM APIs) and find the one with `menuOpen`, so the prototype does that currently.

For now the HLD mentions both options. The prototype also has my original code that walks the DOM still there but commented out, for comparison. Not sure if we need to nail down all the details here (in this HLD) or not.

msmithNI replied:
As mentioned above I think we can nail down the details more here with the actual implementation (separate from the HLD), so marking Resolved.

rajsite commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
```diff
@@ -147,7 +145,37 @@ abstract class TableColumn<TCellData extends TableRecord = TableRecord, TColumnC
 _Note: The `TableColumn` class may be updated to support other features not covered in this HLD such as sorting and grouping._
 
-Given the above class, a series of column elements to handle basic use cases can be written within Nimble. For example, the `TableColumn` implementation we could create for rendering data as a read-only `NimbleTextField` could look like this:
+### `BaseCellElement<>`
```
I think using the name element is a bit redundant, I like the term cell view as used by `cellViewTag`. Also our TableColumn base class isn't using `Base` in the name so to be consistent at least for column authors I'd leave it off.

So I'd suggest the name `TableCellView` (it is close to TableCell but I think it's actually not a bad distinction of TableCell vs TableCellView)

msmithNI replied:
I'm definitely up for alternate naming proposals. I think the main change I wanted to incorporate from your initial pseudocode was having `Cell` in the name to make it clear that the element will used repeatedly in individual cells - I didn't think something like `[...]ColumnView[...]` conveyed that.

msmithNI replied:
Updated to `TableCellView`, and updated the examples in the table columns HLD from `*CellElement` to `*CellView`.

msmithNI commented:
@rajsite @atmgrifter00 @jattasNI @mollykreis this should be ready to re-review now, with the custom element changes we discussed. There'll probably need some more changes needed to naming of things like `cellViewTag` and `BaseCellElement` but I think it's ready for another round of feedback at least.

rajsite commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
```diff
@@ -250,38 +313,45 @@ public class TableColumnPositiveNegativeNumber extends TableColumn<TableColumnPo
+class PositiveNegativeNumberCellElement extends BaseCellElement<TableColumnPositiveNegativeNumberCellData> {
+    @volatile
+    public get textFieldCssClass(): string {
+        return this.cellRecord.value > 0 ? 'good' : 'bad';
```
For this to work the fields of the cellRecord objects themselves need to be observable I think. Although if we always update the full cellRecord object reference (ie activate the setter for the cellRecord property) then it should be fine. We just can't update the fields directly, we have to create a new object that is assigned to the setter of cellRecord (so it passes a diff check).

atmgrifter00 replied:
Pretty sure we will never update the fields of an existing cellRecord.

mollykreis commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,119 @@
+
+In this case, `document.activeElement` will be the Nimble `Table`, and `table.shadowRoot.activeElement` will be a Nimble `TableRow`. We can recursively look at the active element's `shadowRoot.activeElement`, starting from the TableRow, and stop when we reach a `BaseCellElement` (the custom element in table cells implemented by column plugin authors) or `null` (see [prototype](https://github.com/ ...
+
+If we found a focused `BaseCellElement`, then we will call the new API `onBeforeBlur()` on it.  
```
I'm struggling to come up with a good alternative, but I'm not a huge fan of `onBeforeBlur()` for this function. At least to me, `onBeforeBlur()` means that the element is about to be blurred so I should take any actions I need to before that blur occurs. However, what is really happening is that I am being told to blur the element myself.

What about something like `blurCell`?

rajsite replied:
I think I see the direction you were going. `blurCell` is like an action, you are telling it to blur. Like the focus method, or the actual blur method lol.

In that vein one option is we actually tell clients that to respond to blur behaviors you need to override the blur method. I think that kinda maybe even makes sense. Like calling the blur method on an input element with an uncommited value commits the value and then removes keyboard focus. Can try it on the mdn page adding a change listener in devtools: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/blur

The other web component lifecycle behaviors are named `*Callback` (connectedCallback, disconnectedCallback, attributeChangedCallback). Maybe a `*TableCallback` to emphasize the lifecycle nature of it. And I like the direction of "what is the table doing" which aligns more with the lifecycle nature, i.e. connected/disconnected/attributeChanged are what is the DOM API doing while `*TableCallback` is what is the table doing. Maybe `beforeRecycleTableCallback`, `beforeBlurTableCallback`?

I think I kind of like the hey we will call `blur()` on you when we want to remove focus so commit your thing at that point approach. I think the behavior that blur on native text fields commited the value and caused it to lose focus, which sounds like exactly what we want, is making me lean to that direction. But if we want more control, i.e. the need to pass different parameters then I'd lean to the `*TableCallback` approach.

rajsite replied:
Another thought that maybe overriding blur (and calling super) answers. What should have focus after the blur?
The contract of the blur method is that it is the viewport that has focus:
![image](https://user-images.githubusercontent.com/1588923/224858613-9c5a1150-33d8-463c-9224-a7b8e2998568.png)
When you page down in a keyboard I'd expect a new cell cell to be focused, not for the table to lose focus. So maybe overriding blur is the wrong behavior and the name blur is a bit misleading / unaligned with the spec definition of blur.

And I think I remember that this is more precise than recycling. This is a callback specific to a cell that is about to be recycled that has focus losing focus and then being recycled. Many other cells will be recycled at the same time but this is specifcally a focused cell about to be recycled.

Option 3: `beforeFocusChangeTableCallback`. It's not a "blur" because focus isn't going to the viewport, it's changing to something else in the table (I'm assuming based on the above discussion once keyboard interactions are implemented). It's different from other lifecycle `*Callback` because it happens before not after.

rajsite replied:
This is actually getting really close to the naming from one of your original proposals of `onBeforeFocusedCellRecycled` lol

Option 4: beforeFocusedCellRecycleCallback, beforeFocusedCellRecycle, beforeFocusedRecycle, beforeFocusedRecycleCallback, focusedRecycleCallback, focusedRecycleTableCallback

I think I like focusedRecycleCallback the most and leaves room for a recycleCallback if needed in the future...

msmithNI replied:
Updated to `focusedRecycleCallback` naming.  
I agree with keeping `blur` out of the name since we won't be the ones calling `blur` ourselves, the column plugins will.

mollykreis commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
```diff
@@ -250,38 +313,45 @@ public class TableColumnPositiveNegativeNumber extends TableColumn<TableColumnPo
 }
+
+class PositiveNegativeNumberCellElement extends BaseCellElement<TableColumnPositiveNegativeNumberCellData> {
+    @volatile
```
nit: It doesn't look like the `textFieldCssClass` getter needs to be marked as `@volatile`. From my understanding, `@volatile` is needed when branching logic changes which observable properties are accessed, which is not the case in this example.

msmithNI replied:
Updated to remove `@volatile` from this example and the one above it. As you said, only the 1st example should need it as it has a branch between using `value` and `placeholder`.

atmgrifter00 commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
```diff
@@ -238,7 +301,7 @@ const isPositive = (value: number): bool => {
 }
 
-public class TableColumnPositiveNegativeNumber extends TableColumn<TableColumnPositiveNegativeNumberCellData> {
+public class TableColumnPositiveNegativeNumber extends TableColumn {
```
Reminder to update the implementation below to use the new custom element tag for its `cellViewTag` override.

msmithNI replied:
Updated

atmgrifter00 commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
```diff
@@ -250,38 +313,45 @@ public class TableColumnPositiveNegativeNumber extends TableColumn<TableColumnPo
 
-Finally, here is a column element that allows a user to register a callback for a click event on a button inside the cell template:
+Finally, here is a column element that allows a user to register a callback for a click event on a button inside the cell template:  
+**TODO: Update the following example based on custom-element-in-cell changes**
```
Here's a comment to use to track this TODO :).

msmithNI replied:
Updated the last example

msmithNI commented on packages/nimble-components/src/table/specs/table-columns-hld.md:
````diff
@@ -147,7 +145,37 @@ abstract class TableColumn<TCellData extends TableRecord = TableRecord, TColumnC
+}
+```
+
+Requiring column plugins to create custom elements for use in the table cells has several implementations:
````
Updated

msmithNI commented:
@rajsite addressed the open issues on this HLD. Keeping it open for now since you haven't approved it yet / in case you want to take a look again.

mollykreis commented on packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md:
```diff
@@ -0,0 +1,119 @@
+
+When the user scrolls the table, we will:
+
+-   If a table cell contains an active/focused control, we will notify the column plugin that the cell is being recycled via a new API. The general expectation is that editable controls will commit any pending updates, then blur the focused the control.
```
```suggestion
-   If a table cell contains an active/focused control, we will notify the column plugin that the cell is being recycled via a new API. The general expectation is that editable controls will commit any pending updates, then blur the focused control.
```
